kernel/src/net/socket/ip/datagram/bound.rs

```diff
@@ -55,26 +65,39 @@ impl BoundDatagram {
 
     pub fn try_send(
         &self,
-        buf: &[u8],
+        reader: &mut dyn IoVecRead,
         remote: &IpEndpoint,
         _flags: SendRecvFlags,
     ) -> Result<usize> {
-        let result = self.bound_socket.raw_with(|socket: &mut RawUdpSocket| {
-            if socket.payload_send_capacity() < buf.len() {
-                return None;
-            }
-            Some(socket.send_slice(buf, *remote))
-        });
-        match result {
-            Some(Ok(())) => Ok(buf.len()),
-            Some(Err(SendError::BufferFull)) => {
-                return_errno_with_message!(Errno::EAGAIN, "the send buffer is full")
+        let reader_len = reader.len();
+
+        self.bound_socket.raw_with(|socket: &mut RawUdpSocket| {
+            if socket.payload_send_capacity() < reader_len {
+                return_errno_with_message!(Errno::EMSGSIZE, "the message is too large");
             }
-            Some(Err(SendError::Unaddressable)) => {
-                return_errno_with_message!(Errno::EINVAL, "the destination address is invalid")
+
+            let res = socket.send_with(reader_len, *remote, |socket_buffer| {
+                match reader.read_to(&mut VmWriter::from(socket_buffer)) {
+                    Ok(len) => len,
+                    Err(_) => {
+                        // FIXME: If copy failed, we should not send any packet.
+                        // But current API seems not support this behavior.
```

Comment:
This FIXME is fine as it is hard to be fixed without changing smoltcp.

But I think there is a tricky difference between UDP's (a message-oriented protocol) `send` and TCP's (a stream-oriented protocol) `send` that is ignored in this PR.

My _guess_ about Linux's behavior is this:
 - If we cannot copy the entire UDP message into the buffer, we send nothing, but fail with `EFAULT`.
 - If we can copy part (at least one byte) of a TCP message into the buffer, we send that part and return the number of bytes in the buffer.

Could you please look into this behavior and consider it in the API design?

Reply:
From my above comment at https://github.com/asterinas/asterinas/pull/1294#discussion_r1751578615. I check the Linux behavior, any failed copy will return EFAULT, it holds for both TCP and UDP. That is to say, even only one io_vec in all io_vecs points to invalid memory, the syscall should return EFAULT.

Reply:
OK, thanks again for your experiments. As mentioned in https://github.com/asterinas/asterinas/pull/1294#discussion_r1751921866, I think it's because Linux kernel maintainers are a bit lazy to create two different modes for copying to the socket buffer, so they just assume all protocols are message-oriented.

This is reasonable. Treating a stream-oriented protocol as message-oriented doesn't do much harm (we don't send anything, and we tell userspace that nothing was sent, which is all fine), but if we do the opposite, we run the risk of sending a truncated message (which is not what userspace really wants).